utils/s2n_safety.h

```diff
@@ -93,6 +93,11 @@
  */
 #define ENSURE_NONNULL( x )                          ENSURE((x) != NULL, S2N_ERR_NULL)
 
+/**
+ * Ensures the `condition` is `true`, otherwise the function will `BAIL_POSIX` with a `S2N_ERR_SAFETY` error
+ */
+#define ENSURE_POSTCONDITION( condition )            __S2N_ENSURE((condition), BAIL_POSIX(S2N_ERR_SAFETY))
```

Comment:
Is the macro actually needed? Wouldn't `ENSURE_POSIX(condition, S2N_ERR_SAFETY)` work instead? I ask because I'd prefer the list of `ENSURE_*` macros not get any lengthier than it already is. :)

If we do decide to keep it, could we define it instead as:

```c
/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL` with a `S2N_ERR_SAFETY` error
 */
#define ENSURE_POSTCONDITION( condition )            ENSURE((condition), S2N_ERR_SAFETY)

/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL_POSIX` with a `S2N_ERR_SAFETY` error
 */
#define ENSURE_POSIX_POSTCONDITION( condition )            ENSURE_POSIX((condition), S2N_ERR_SAFETY)
```

This will make the migration to `s2n_result` easier :)

Reply:
It provides some semantic information, in that post-conditions are contractural guarantees, while ensures are internal checks.  But the benefit may not be worth the confusion.

Reply:
@camshaft I implemented as you suggested, so we can agree in a middle ground :) I do think the semantic information is valid and it makes clear that we have a contractual guarantee when adding pre- and post-conditions to a function.

Reply:
I'm fine with keeping it. Do you envision adding `ENSURE_PRECONDITION` as well? Also I wonder if we want to drop the `ENSURE_` prefix and just do `POSTCONDITION`?

Reply:
@camshaft We have now `S2N_PRECONDITION`. So we could just name it `S2N_POSTCONDITION`. What do you think it's best?

Reply:
Ah right! [This one](https://github.com/awslabs/s2n/blob/master/error/s2n_errno.h#L272)... Yeah, it really doesn't belong in there.

Since the macros from both `s2n_errno.h` and `s2n_safety.h` are private, I don't think we need the `S2N_` prefix. So my vote would be:

```c
// s2n_safety.h

/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL` with a `S2N_ERR_SAFETY` error
 */
#define PRECONDITION( condition )                 ENSURE((condition), S2N_ERR_PRECONDITION_VIOLATION)

/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL` with a `S2N_ERR_SAFETY` error
 */
#define POSTCONDITION( condition )                ENSURE((condition), S2N_ERR_POSTCONDITION_VIOLATION)

/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL_POSIX` with a `S2N_ERR_SAFETY` error
 */
#define PRECONDITION_POSIX( condition )           ENSURE_POSIX((condition), S2N_ERR_PRECONDITION_VIOLATION)

/**
 * Ensures the `condition` is `true`, otherwise the function will `BAIL_POSIX` with a `S2N_ERR_SAFETY` error
 */
#define POSTCONDITION_POSIX( condition )          ENSURE_POSIX((condition), S2N_ERR_POSTCONDITION_VIOLATION)
```

Reply:
Nice!! I implemented as you suggested and I also added the `__TIMING_CONTRACTS__` guard.

Reply:
You can actually remove the timing contracts noop definitions entirely. That's already handled when `s2n_ensure.h` is replaced in sidetrail tests.

Reply:
Awesome, I didn't know that. I removed the `__TIMING_CONTRACTS__` then.